src/main/java/io/supertokens/thirdparty/ThirdParty.java

```diff
@@ -250,35 +247,75 @@ private static SignInUpResponse signInUpHelper(TenantIdentifierWithStorage tenan
+                                        loginMethod.thirdParty.userId.equals(thirdPartyUserId)) {
+                                    lM1 = loginMethod;
+                                    break;
+                                }
+                            }
+
+                            if (lM1 == null) {
+                                throw new IllegalStateException("Should never come here");
+                            }
+
+                            if (!email.equals(lM1.email)) {
+                                // before updating the email, we must check for if another primary user has the same
+                                // email, and if they do, then we do not allow the update.
+                                if (userFromDb1.isPrimaryUser) {
+                                    for (String tenantId : userFromDb1.tenantIds) {
+                                        AuthRecipeUserInfo[] userBasedOnEmail =
+                                                authRecipeStorage.listPrimaryUsersByEmail_Transaction(
+                                                        appIdentifier, con, email
+                                                );
+                                        for (AuthRecipeUserInfo userWithSameEmail : userBasedOnEmail) {
+                                            if (!userWithSameEmail.tenantIds.contains(tenantId)) {
+                                                continue;
+                                            }
+                                            if (userWithSameEmail.isPrimaryUser &&
+                                                    !userWithSameEmail.getSupertokensUserId().equals(userFromDb1.getSupertokensUserId())) {
+                                                throw new StorageTransactionLogicException(
+                                                        new EmailChangeNotAllowedException());
+                                            }
+                                        }
                                     }
                                 }
+                                storage.updateUserEmail_Transaction(appIdentifier, con,
+                                        thirdPartyId, thirdPartyUserId, email);
                             }
-                        }
-                        storage.updateUserEmail_Transaction(appIdentifier, con,
-                                thirdPartyId, thirdPartyUserId, email);
-                    }
 
-                    storage.commitTransaction(con);
-                    return null;
-                });
+                            storage.commitTransaction(con);
+                            return null;
+                        });
+                    }
+                }
 
                 AuthRecipeUserInfo user = getUser(tenantIdentifierWithStorage, thirdPartyId, thirdPartyUserId);
                 return new SignInUpResponse(false, user);
```

Comment:
the try catch below is incorrectly scoped

Reply:
Done